Gk-deploy failing on Ubuntu

I'm trying to install Gluster daemon hyperconverged on a completely fresh Kubernetes cluster on 3 Ubuntu 18.04 nodes, which are also the only nodes in the Kubernetes cluster (all masters).

I've done the following:
1. Configured (but not initialized) the /dev/sdb drive on each host for exclusive Gluster use.
2. Installed Gluster (client only) using apt install glusterfs-client (version 4.1.6)
3. Made sure the module and firewall requirements listed here are loaded: https://github.com/gluster/gluster-kubernetes/blob/master/docs/setup-guide.md
4. Created topology file 
5. Run ./gk-deploy -gv

The end result (see verbose output in attachment)
[Gk-deploy verbose output.txt](https://github.com/gluster/gluster-kubernetes/files/2617346/Gk-deploy.verbose.output.txt)
 is 3 glusterfs pods showing this in their kubectl describe output:

`Events:
  Type     Reason     Age                 From                  Message
  ----     ------     ----                ----                  -------
  Normal   Pulled     19m                 kubelet, vva-er-k8s0  Container image "gluster/gluster-centos:latest" already present on machine
  Normal   Created    19m                 kubelet, vva-er-k8s0  Created container
  Normal   Started    19m                 kubelet, vva-er-k8s0  Started container
  Warning  Unhealthy  14m (x11 over 18m)  kubelet, vva-er-k8s0  Readiness probe failed: /usr/local/bin/status-probe.sh
failed check: systemctl -q is-active gluster-blockd.service
  Warning  Unhealthy  4m10s (x35 over 18m)  kubelet, vva-er-k8s0  Liveness probe failed: /usr/local/bin/status-probe.sh
failed check: systemctl -q is-active gluster-blockd.service`

If I enter one of the containers, I can see the following:

input: systemctl -q

output:  
`glusterd.service                                                 loaded active     running   GlusterFS, a clustered file-system server`

input: cat /var/log/glusterfs/glusterd.log

output:
`[root@vva-er-k8s0 glusterfs]# cat glusterd.log
[2018-11-26 20:49:11.437520] I [MSGID: 100030] [glusterfsd.c:2741:main] 0-/usr/sbin/glusterd: Started running /usr/sbin/glusterd version 4.1.5 (args: /usr/sbin/glusterd -p /var/run/glusterd.pid --log-level INFO)
[2018-11-26 20:49:11.441727] I [MSGID: 106478] [glusterd.c:1423:init] 0-management: Maximum allowed open file descriptors set to 65536
[2018-11-26 20:49:11.441752] I [MSGID: 106479] [glusterd.c:1481:init] 0-management: Using /var/lib/glusterd as working directory
[2018-11-26 20:49:11.441759] I [MSGID: 106479] [glusterd.c:1486:init] 0-management: Using /var/run/gluster as pid file working directory
[2018-11-26 20:49:11.446193] W [MSGID: 103071] [rdma.c:4629:__gf_rdma_ctx_create] 0-rpc-transport/rdma: rdma_cm event channel creation failed [No such device]
[2018-11-26 20:49:11.446212] W [MSGID: 103055] [rdma.c:4938:init] 0-rdma.management: Failed to initialize IB Device
[2018-11-26 20:49:11.446220] W [rpc-transport.c:351:rpc_transport_load] 0-rpc-transport: 'rdma' initialization failed
[2018-11-26 20:49:11.446289] W [rpcsvc.c:1781:rpcsvc_create_listener] 0-rpc-service: cannot create listener, initing the transport failed
[2018-11-26 20:49:11.446299] E [MSGID: 106244] [glusterd.c:1764:init] 0-management: creation of 1 listeners failed, continuing with succeeded transport
[2018-11-26 20:49:12.457991] E [MSGID: 101032] [store.c:441:gf_store_handle_retrieve] 0-: Path corresponding to /var/lib/glusterd/glusterd.info. [No such file or directory]
[2018-11-26 20:49:12.458038] E [MSGID: 101032] [store.c:441:gf_store_handle_retrieve] 0-: Path corresponding to /var/lib/glusterd/glusterd.info. [No such file or directory]
[2018-11-26 20:49:12.458040] I [MSGID: 106514] [glusterd-store.c:2262:glusterd_restore_op_version] 0-management: Detected new install. Setting op-version to maximum : 40100
[2018-11-26 20:49:12.468372] I [MSGID: 106194] [glusterd-store.c:3849:glusterd_store_retrieve_missed_snaps_list] 0-management: No missed snaps list.
Final graph:
+------------------------------------------------------------------------------+
  1: volume management
  2:     type mgmt/glusterd
  3:     option rpc-auth.auth-glusterfs on
  4:     option rpc-auth.auth-unix on
  5:     option rpc-auth.auth-null on
  6:     option rpc-auth-allow-insecure on
  7:     option transport.listen-backlog 10
  8:     option event-threads 1
  9:     option ping-timeout 0
 10:     option transport.socket.read-fail-log off
 11:     option transport.socket.keepalive-interval 2
 12:     option transport.socket.keepalive-time 10
 13:     option transport-type rdma
 14:     option working-directory /var/lib/glusterd
 15: end-volume
 16:
+------------------------------------------------------------------------------+
[2018-11-26 20:49:12.468685] I [MSGID: 101190] [event-epoll.c:617:event_dispatch_epoll_worker] 0-epoll: Started thread with index 1
[2018-11-26 21:10:59.076690] I [MSGID: 100030] [glusterfsd.c:2741:main] 0-/usr/sbin/glusterd: Started running /usr/sbin/glusterd version 4.1.5 (args: /usr/sbin/glusterd -p /var/run/glusterd.pid --log-level INFO)
[2018-11-26 21:10:59.080303] I [MSGID: 106478] [glusterd.c:1423:init] 0-management: Maximum allowed open file descriptors set to 65536
[2018-11-26 21:10:59.080327] I [MSGID: 106479] [glusterd.c:1481:init] 0-management: Using /var/lib/glusterd as working directory
[2018-11-26 21:10:59.080335] I [MSGID: 106479] [glusterd.c:1486:init] 0-management: Using /var/run/gluster as pid file working directory
[2018-11-26 21:10:59.084469] W [MSGID: 103071] [rdma.c:4629:__gf_rdma_ctx_create] 0-rpc-transport/rdma: rdma_cm event channel creation failed [No such device]
[2018-11-26 21:10:59.084488] W [MSGID: 103055] [rdma.c:4938:init] 0-rdma.management: Failed to initialize IB Device
[2018-11-26 21:10:59.084496] W [rpc-transport.c:351:rpc_transport_load] 0-rpc-transport: 'rdma' initialization failed
[2018-11-26 21:10:59.084565] W [rpcsvc.c:1781:rpcsvc_create_listener] 0-rpc-service: cannot create listener, initing the transport failed
[2018-11-26 21:10:59.084575] E [MSGID: 106244] [glusterd.c:1764:init] 0-management: creation of 1 listeners failed, continuing with succeeded transport
[2018-11-26 21:11:00.027597] E [MSGID: 101032] [store.c:441:gf_store_handle_retrieve] 0-: Path corresponding to /var/lib/glusterd/glusterd.info. [No such file or directory]
[2018-11-26 21:11:00.027644] E [MSGID: 101032] [store.c:441:gf_store_handle_retrieve] 0-: Path corresponding to /var/lib/glusterd/glusterd.info. [No such file or directory]
[2018-11-26 21:11:00.027646] I [MSGID: 106514] [glusterd-store.c:2262:glusterd_restore_op_version] 0-management: Detected new install. Setting op-version to maximum : 40100
[2018-11-26 21:11:00.034982] I [MSGID: 106194] [glusterd-store.c:3849:glusterd_store_retrieve_missed_snaps_list] 0-management: No missed snaps list.
Final graph:
+------------------------------------------------------------------------------+
  1: volume management
  2:     type mgmt/glusterd
  3:     option rpc-auth.auth-glusterfs on
  4:     option rpc-auth.auth-unix on
  5:     option rpc-auth.auth-null on
  6:     option rpc-auth-allow-insecure on
  7:     option transport.listen-backlog 10
  8:     option event-threads 1
  9:     option ping-timeout 0
 10:     option transport.socket.read-fail-log off
 11:     option transport.socket.keepalive-interval 2
 12:     option transport.socket.keepalive-time 10
 13:     option transport-type rdma
 14:     option working-directory /var/lib/glusterd
 15: end-volume
 16:
+------------------------------------------------------------------------------+
[2018-11-26 21:11:00.035295] I [MSGID: 101190] [event-epoll.c:617:event_dispatch_epoll_worker] 0-epoll: Started thread with index 1
`

Every few minutes, docker kills and recreates the gluster container.

What am I missing here?

I got this from inside the of one of the glusterfs pods:

```
[root@glusterfs2 /]# tcmu-runner -dd
The logdir option from the tcmu.conf will be ignored
2019-01-16 03:16:55.012 288 [INFO] dyn_config_start:409: Inotify is watching "/etc/tcmu/tcmu.conf", wd: 1, mask: IN_ALL_EVENTS
2019-01-16 03:16:55.012 288 [DEBUG] main:1018: handler path: /usr/lib64/tcmu-runner
2019-01-16 03:16:55.013 288 [INFO] load_our_module:498: no modules directory '/lib/modules/4.15.0-43-generic', checking module target_core_user entry in '/sys/modules/'
2019-01-16 03:16:55.013 288 [ERROR] load_our_module:503: stat() on '/sys/module/target_core_user' failed: No such file or directory
2019-01-16 03:16:55.013 288 [ERROR] main:1022: couldn't load module
```

Here's another super-odd thing. If I run the command like crazy (several times in a row) I get different messages: 

```
[root@glusterfs2 /]# tcmu-runner -d
The logdir option from the tcmu.conf will be ignored
2019-01-16 03:19:57.426 536 [DEBUG] main:1018: handler path: /usr/lib64/tcmu-runner
2019-01-16 03:19:57.426 536 [INFO] dyn_config_start:409: Inotify is watching "/etc/tcmu/tcmu.conf", wd: 1, mask: IN_ALL_EVENTS
2019-01-16 03:19:57.426 536 [INFO] load_our_module:498: no modules directory '/lib/modules/4.15.0-43-generic', checking module target_core_user entry in '/sys/modules/'
2019-01-16 03:19:57.426 536 [ERROR] load_our_module:503: stat() on '/sys/module/target_core_user' failed: No such file or directory
2019-01-16 03:19:57.426 536 [ERROR] main:1022: couldn't load module
[root@glusterfs2 /]# tcmu-runner -d
The logdir option from the tcmu.conf will be ignored
2019-01-16 03:19:57.667 539 [INFO] dyn_config_start:409: Inotify is watching "/etc/tcmu/tcmu.conf", wd: 1, mask: IN_ALL_EVENTS
2019-01-16 03:19:57.667 539 [DEBUG] main:1018: handler path: /usr/lib64/tcmu-runner
2019-01-16 03:19:57.667 539 [INFO] load_our_module:498: no modules directory '/lib/modules/4.15.0-43-generic', checking module target_core_user entry in '/sys/modules/'
[root@glusterfs2 /]# tcmu-runner -d
The logdir option from the tcmu.conf will be ignored
2019-01-16 03:19:57.906 542 [DEBUG] main:1018: handler path: /usr/lib64/tcmu-runner
2019-01-16 03:19:57.906 542 [INFO] dyn_config_start:409: Inotify is watching "/etc/tcmu/tcmu.conf", wd: 1, mask: IN_ALL_EVENTS
[root@glusterfs2 /]# tcmu-runner -d
The logdir option from the tcmu.conf will be ignored
2019-01-16 03:19:58.161 545 [DEBUG] main:1018: handler path: /usr/lib64/tcmu-runner
2019-01-16 03:19:58.161 545 [INFO] dyn_config_start:409: Inotify is watching "/etc/tcmu/tcmu.conf", wd: 1, mask: IN_ALL_EVENTS
2019-01-16 03:19:58.162 545 [INFO] load_our_module:498: no modules directory '/lib/modules/4.15.0-43-generic', checking module target_core_user entry in '/sys/modules/'
2019-01-16 03:19:58.162 545 [ERROR] load_our_module:503: stat() on '/sys/module/target_core_user' failed: No such file or directory
2019-01-16 03:19:58.162 545 [ERROR] main:1022: couldn't load module
[root@glusterfs2 /]# tcmu-runner -d
The logdir option from the tcmu.conf will be ignored
Inotify is watching "/etc/tcmu/tcmu.conf", wd: 1, mask: IN_ALL_EVENTS
2019-01-16 03:19:58.417 548 [DEBUG] main:1018: handler path: /usr/lib64/tcmu-runner
[root@glusterfs2 /]# tcmu-runner -d
The logdir option from the tcmu.conf will be ignored
Inotify is watching "/etc/tcmu/tcmu.conf", wd: 1, mask: IN_ALL_EVENTS
2019-01-16 03:19:58.702 551 [DEBUG] main:1018: handler path: /usr/lib64/tcmu-runner
[root@glusterfs2 /]# tcmu-runner -d
The logdir option from the tcmu.conf will be ignored
[root@glusterfs2 /]# tcmu-runner -d
The logdir option from the tcmu.conf will be ignored
2019-01-16 03:19:59.179 557 [INFO] dyn_config_start:409: Inotify is watching "/etc/tcmu/tcmu.conf", wd: 1, mask: IN_ALL_EVENTS
2019-01-16 03:19:59.180 557 [DEBUG] main:1018: handler path: /usr/lib64/tcmu-runner
```

On Tue, Jan 15, 2019 at 07:17:31PM -0800, Renich Bon Ciric wrote:
> I got this from inside the of one of the glusterfs pods:
> 
> ```
> [root@glusterfs2 /]# tcmu-runner -dd
> The logdir option from the tcmu.conf will be ignored
> 2019-01-16 03:16:55.012 288 [INFO] dyn_config_start:409: Inotify is watching "/etc/tcmu/tcmu.conf", wd: 1, mask: IN_ALL_EVENTS
> 2019-01-16 03:16:55.012 288 [DEBUG] main:1018: handler path: /usr/lib64/tcmu-runner
> 2019-01-16 03:16:55.013 288 [INFO] load_our_module:498: no modules directory '/lib/modules/4.15.0-43-generic', checking module target_core_user entry in '/sys/modules/'
> 2019-01-16 03:16:55.013 288 [ERROR] load_our_module:503: stat() on '/sys/module/target_core_user' failed: No such file or directory
> 2019-01-16 03:16:55.013 288 [ERROR] main:1022: couldn't load module
> ```

Is /lib not a symlink to /usr/lib on Ubuntu (whatever version)? The
kernel modules that needs to get loaded (target_core_user and others)
need to be available in the container image. We do this by bind-mounting
/usr/lib/modules from the host to /usr/lib/modules in the container. But
if /lib/modules is needed on Ubuntu, we could consider changing it in
the templates.

Can you modify the entry for 'volumes:' and 'volumeMounts:' to use
"/lib/modules" in the glusterfs-server pods?

--- a/deploy/kube-templates/glusterfs-daemonset.yaml
+++ b/deploy/kube-templates/glusterfs-daemonset.yaml
@@ -67,7 +67,7 @@ spec:
           mountPath: "/etc/ssl"
           readOnly: true
         - name: kernel-modules
-          mountPath: "/usr/lib/modules"
+          mountPath: "/lib/modules"
           readOnly: true
         securityContext:
           capabilities: {}
@@ -131,4 +131,4 @@ spec:
           path: "/etc/ssl"
       - name: kernel-modules
         hostPath:
-          path: "/usr/lib/modules"
+          path: "/lib/modules"

--- a/deploy/ocp-templates/glusterfs-template.yaml
+++ b/deploy/ocp-templates/glusterfs-template.yaml
@@ -77,7 +77,7 @@ objects:
             mountPath: "/etc/ssl"
             readOnly: true
           - name: kernel-modules
-            mountPath: "/usr/lib/modules"
+            mountPath: "/lib/modules"
             readOnly: true
           securityContext:
             capabilities: {}
@@ -143,7 +143,7 @@ objects:
             path: "/etc/ssl"
         - name: kernel-modules
           hostPath:
-            path: "/usr/lib/modules"
+            path: "/lib/modules"
         restartPolicy: Always
         terminationGracePeriodSeconds: 30
         dnsPolicy: ClusterFirst

